editor/plugins/material_editor_plugin.cpp

```diff
@@ -35,44 +35,87 @@
-
-			first_enter = false;
+void MaterialEditor::_gui_input(Ref<InputEvent> p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid() && mm->get_button_mask() & MOUSE_BUTTON_MASK_LEFT) {
+		rot_x -= mm->get_relative().y * 0.01;
+		rot_y -= mm->get_relative().x * 0.01;
+		if (rot_x < -Math_PI / 2) {
+			rot_x = -Math_PI / 2;
+		} else if (rot_x > Math_PI / 2) {
+			rot_x = Math_PI / 2;
 		}
+		_update_rotation();
 	}
+}
 
-	if (p_what == NOTIFICATION_DRAW) {
-		Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
-		Size2 size = get_size();
-
-		draw_texture_rect(checkerboard, Rect2(Point2(), size), true);
+void MaterialEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			if (first_enter) {
+				light_1_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight1", "EditorIcons"));
+				light_1_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight1Off", "EditorIcons"));
+				light_2_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight2", "EditorIcons"));
+				light_2_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight2Off", "EditorIcons"));
+
+				sphere_switch->set_normal_texture(get_theme_icon("MaterialPreviewSphereOff", "EditorIcons"));
+				sphere_switch->set_pressed_texture(get_theme_icon("MaterialPreviewSphere", "EditorIcons"));
+				box_switch->set_normal_texture(get_theme_icon("MaterialPreviewCubeOff", "EditorIcons"));
+				box_switch->set_pressed_texture(get_theme_icon("MaterialPreviewCube", "EditorIcons"));
+				first_enter = false;
+			}
+		} break;
+		case NOTIFICATION_DRAW: {
+			Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
+			Size2 size = get_size();
+
+			draw_texture_rect(checkerboard, Rect2(Point2(), size), true);
+		} break;
 	}
 }
 
+void MaterialEditor::_update_rotation() {
+	Transform3D t;
+	t.basis.rotate(Vector3(0, 1, 0), -rot_y);
+	t.basis.rotate(Vector3(1, 0, 0), -rot_x);
+	rotation->set_transform(t);
+}
+
 void MaterialEditor::edit(Ref<Material> p_material, const Ref<Environment> &p_env) {
 	material = p_material;
 	camera->set_environment(p_env);
+
 	if (!material.is_null()) {
 		sphere_instance->set_material_override(material);
 		box_instance->set_material_override(material);
 	} else {
 		hide();
 	}
+
+	rot_x = Math::deg2rad(-15.0);
+	rot_y = Math::deg2rad(30.0);
+	_update_rotation();
+
+	AABB aabb = box_instance->get_aabb();
+
+	if (sphere_switch->is_pressed()) {
+		aabb = sphere_instance->get_aabb();
+	}
+
+	Vector3 ofs = aabb.position + aabb.size * 0.5;
+	float m = aabb.get_longest_axis_size();
+	if (m != 0) {
+		m = 1.0 / m;
+		m *= 0.5;
+		Transform3D xform;
+		xform.basis.scale(Vector3(m, m, m));
+		xform.origin = -xform.basis.xform(ofs); //-ofs*m;
+		//xform.origin.z -= aabb.get_longest_axis_size() * 2;
+		box_instance->set_transform(xform);
+		sphere_instance->set_transform(xform);
+	}
```

Comment:
I don't understand the purpose of these lines. I replaced all 18 of these lines with just this, and it works perfectly fine:

```suggestion
	box_instance->set_transform(Transform3D(Basis().scaled(Vector3(0.25, 0.25, 0.25))));
	sphere_instance->set_transform(Transform3D(Basis().scaled(Vector3(0.375, 0.375, 0.375))));
```